File: app/src/main/java/com/test/murphy/weatherapp/model/WeatherForecast.java
package com.test.murphy.weatherapp.model;

import android.os.Parcel;
import android.os.Parcelable;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Created by wsmurphy on 10/23/17.
 */

public class WeatherForecast implements Parcelable {
    public String location;
    public WeatherConditions[] forecast;

    public WeatherForecast(JSONObject jsonObject) {
        try {
            location = jsonObject.getString("city");

            JSONArray list = jsonObject.getJSONArray("list");

            forecast = new WeatherConditions[list.length()];
            for (int i = 0; i < list.length(); i++) {
                JSONObject item = list.getJSONObject(i);
                WeatherConditions conditions = new WeatherConditions(item);
                forecast[i] = conditions;
            }
        } catch (JSONException je) {
            //TODO: Handle exception
        }

    }

    @Override
    public int describeContents() {
        return 0;
    }

    public void writeToParcel(Parcel out, int flags) {
        out.writeString(location);
        out.writeTypedArray(forecast, 0);
    }

    public static final Parcelable.Creator<WeatherForecast> CREATOR
            = new Parcelable.Creator<WeatherForecast>() {
        public WeatherForecast createFromParcel(Parcel in) {
            return new WeatherForecast(in);
        }

        public WeatherForecast[] newArray(int size) {
            return new WeatherForecast[size];
        }
    };

    private WeatherForecast(Parcel in) {
        location = in.readString();
        forecast = in.createTypedArray(WeatherConditions.CREATOR);
    }
}

File: app/src/main/java/com/test/murphy/weatherapp/model/WeatherForecast.kt
package com.test.murphy.weatherapp.model

import android.os.Parcel
import android.os.Parcelable
import org.json.JSONObject
import java.util.*

/**
 * Created by wsmurphy on 10/23/17.
 */


class WeatherForecast(var location: String, var forecast: Array<WeatherConditions>) : Parcelable {

    constructor(parcel: Parcel) : this(
            parcel.readString(),
            parcel.createTypedArray(WeatherConditions.CREATOR))

    constructor(jsonObject: JSONObject) : this(
            jsonObject.getString("city"),
            WeatherForecast.parseForecastJSON(jsonObject))

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(location)
        parcel.writeTypedArray(forecast, flags)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<WeatherForecast> {
        override fun createFromParcel(parcel: Parcel): WeatherForecast {
            return WeatherForecast(parcel)
        }

        override fun newArray(size: Int): Array<WeatherForecast?> {
            return arrayOfNulls(size)
        }

        fun parseForecastJSON(jsonObject: JSONObject) : Array<WeatherConditions> {
            val list = jsonObject.getJSONArray("list")

            var forecast: ArrayList<WeatherConditions> = ArrayList(list.length())

            for (i in 0 until list.length()) {
                val item = list.getJSONObject(i)
                val conditions = WeatherConditions(item)
                forecast.add(conditions)
            }

            return forecast.toTypedArray()
        }
    }
}
